### PR TITLE
Docker packaging should not su-exec or chmod if already running as UID/GID

### PR DESCRIPTION
**Pull request by michaelkaye**
It's possible to configure docker to set the UID/GID before a container starts.

Some docker based environments, such as openshift, [force containers to run with a specific UID](https://docs.openshift.com/enterprise/3.1/creating_images/guidelines.html#use-uid) and GID and cannot start as 0:0 and drop to a different user to continue execution.

This PR checks whether the environment of the docker container is that of the uid/gid running start.py are the same as the UID/GID specified in the environment.

If this is true, then it does not attempt any chmod / su-exec based operations, as:

- su-exec requires root to function
- chmod will work only if the target uid:gid matches the existing permissions.

so these will not generally succeed if the container is not already running as root.

Existing deployments using root:root to run the container will be unaffected, as their UID/GID will not match.
Existing deployments using another UID/GID to run the container currently do not function as they're unable to use su-exec to change.

If UID != getuid but getuid != root, then running the container will fail with a chmod / su-exec error that it's unable to assume the group, 

I think this could also be expressed as a check for getuid/getgid not being root:root and if that's true, set `ownership = None`, but that would then not attempt to perform the changes, and you would be able to supply a UID/GID but not use them. But happy to have thoughts on whether that would be better.

### Pull Request Checklist

* [X] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [ ] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### CODE REVIEW

File: docker/start.py [hunk 1]
```diff
@@ -122,7 +118,12 @@ def generate_config_from_template(config_dir, config_path, environ, ownership):
             config_dir,
             "--generate-keys",
         ]
-    )
+
+    if ownership is not None:
```
Comment (richvdh) [inline docker/start.py]: please can you update the docstring to say that `ownership` can be `None`, and what that means?

File: docker/start.py [hunk 2]
```diff
@@ -170,13 +168,27 @@ def run_generate_config(environ, ownership):
         "--open-private-ports",
     ]
     # log("running %s" % (args, ))
-    os.execv("/usr/local/bin/python", args)
+
+    if ownership is not None:
```
Comment (richvdh) [inline docker/start.py]: again, docstring please

File: changelog.d/5970.misc [hunk 3]
```diff
@@ -0,0 +1 @@
+Prevent Docker changing UID/GID if already correct.
```
Comment (richvdh) [inline changelog.d/5970.misc]: could you stick this in a `.docker` changelog entry please?

Comment (richvdh) [inline changelog.d/5970.misc]:
```suggestion
Avoid changing UID/GID if they are already correct.
```

**Comment by michaelkaye** [comment]
I've also updated the logging lines so it's clearer when we choose not to run chmod - hopefully this will at least make debugging why odd UID / getuid() mismatches are causing pain when running the container.

File: docker/start.py [hunk 4]
```diff
@@ -42,7 +42,7 @@ def generate_config_from_template(config_dir, config_path, environ, ownership):
         config_path (str): where to put the main config file
         environ (dict): environment dictionary
         ownership (str): "<user>:<group>" string which will be used to set
```
Comment (richvdh) [inline docker/start.py]:
```suggestion
        ownership (str|None): "<user>:<group>" string which will be used to set
```

File: docker/start.py [hunk 5]
```diff
@@ -105,32 +105,32 @@ def generate_config_from_template(config_dir, config_path, environ, ownership):
...
+    subprocess.check_output(args)
 
 
 def run_generate_config(environ, ownership):
     """Run synapse with a --generate-config param to generate a template config file
 
     Args:
         environ (dict): env var dict
-        ownership (str): "userid:groupid" arg for chmod
+        ownership (str): "userid:groupid" arg for chmod. If None, ownership will not change.
```
Comment (richvdh) [inline docker/start.py]:
```suggestion
        ownership (str|None): "userid:groupid" arg for chmod. If None, ownership will not change.
```

**Pull request merged**